# Exploração de Dados — Projeto de Machine Learning

Este notebook deve ser usado para a **primeira análise do dataset** escolhido pelo grupo.

O objetivo é entender a base antes de treinar modelos:

- Qual é o problema?
- Quais colunas existem?
- Qual é a variável-alvo?
- Existem dados faltantes?

> Ao final, o grupo deve conseguir explicar o que entendeu da base e quais decisões pretende tomar antes da modelagem.

## 1. Identificação do grupo

Preencha as informações abaixo:

- **Grupo:** 03
- **Integrantes:** Douglas Lacerda, Guilherme Trajano, Ian Teixeira, Luiz Daros e Márcio Medeiros
- **Projeto escolhido:** Email Phishing Dataset
- **Link do dataset no Kaggle:** [dataset](https://www.kaggle.com/datasets/ethancratchley/email-phishing-dataset?authuser=1)
- **Pergunta principal do projeto:** Quais fatores influenciam um email ser classificado como spam?

In [1]:
# Bibliotecas principais
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

# Configurações de visualização
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', 100)

## 2. Carregamento da base

Você pode carregar a base de duas formas:

1. Fazendo upload manual do arquivo `.csv` no Colab;
2. Montando o Google Drive e lendo o arquivo de lá.

Use a opção que for mais simples para o grupo.

In [ ]:
df = pd.read_csv("../email_phishing_data.csv")

# Visualizar as primeiras linhas
df.head()

,num_words,num_unique_words,num_stopwords,num_links,num_unique_domains,num_email_addresses,num_spelling_errors,num_urgent_keywords,label
0,140,94,52,0,0,0,0,0,0
1,5,5,1,0,0,0,0,0,0
2,34,32,15,0,0,0,0,0,0
3,6,6,2,0,0,0,0,0,0
4,9,9,2,0,0,0,0,0,0


## 3. Visão geral da base

Nesta etapa, o grupo deve observar o tamanho da base, os tipos das variáveis e a estrutura geral do dataset.

In [5]:
# Quantidade de linhas e colunas
print(f'Linhas: {df.shape[0]}')
print(f'Colunas: {df.shape[1]}')

Linhas: 524846
Colunas: 9


In [6]:
# Nomes das colunas
list(df.columns)

['num_words',
 'num_unique_words',
 'num_stopwords',
 'num_links',
 'num_unique_domains',
 'num_email_addresses',
 'num_spelling_errors',
 'num_urgent_keywords',
 'label']

In [ ]:
# Informações gerais: tipos de dados e valores não nulos
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 524846 entries, 0 to 524845
Data columns (total 9 columns):
 #   Column               Non-Null Count   Dtype
---  ------               --------------   -----
 0   num_words            524846 non-null  int64
 1   num_unique_words     524846 non-null  int64
 2   num_stopwords        524846 non-null  int64
 3   num_links            524846 non-null  int64
 4   num_unique_domains   524846 non-null  int64
 5   num_email_addresses  524846 non-null  int64
 6   num_spelling_errors  524846 non-null  int64
 7   num_urgent_keywords  524846 non-null  int64
 8   label                524846 non-null  int64
dtypes: int64(9)
memory usage: 36.0 MB


In [ ]:
# Amostra aleatória da base
df.sample(5, random_state=42)

,num_words,num_unique_words,num_stopwords,num_links,num_unique_domains,num_email_addresses,num_spelling_errors,num_urgent_keywords,label
371362,164,108,62,0,0,0,5,0,0
74083,349,181,42,0,0,1,38,0,0
262179,24,22,8,0,0,0,1,0,0
65292,123,84,49,0,0,0,4,0,0
326913,38,29,7,0,0,0,2,0,0


### Escreva aqui sua interpretação

Responda:

- A) Quantas linhas e colunas existem?
- B) Cada linha representa o quê?
- C) Quais colunas parecem ser mais importantes?
- D) Existem colunas que parecem ser identificadores e talvez não ajudem no modelo?

**Resposta do grupo:**

* A) Existem 524846 linhas e 9 colunas no dataset.
* B) Cada linha representa uma instância da base de dados, que nesse caso é um email
* C) A feature "num_urgent_keywords", que se refere à presença de palavras "urgentes" em um email  (como "urgent", "verify" e "update")  parece estar bastante relacionada com a classificação do email como phishing ou não, visto que normalmente emails suspeitos solicitam ações "urgentes" ao usuário/leitor. Além disso, as features "num_unique_words" e "num_stopwords", que se referem a presença de palavras únicas e conectivos no texto do email também podem indicar um email suspeito ou não, visto que um texto mais complexo e elaborado tem uma quantidade de "unique_words" e "stopwords" muito maior e isso normalmente se enquadraria como um email normal(não suspeito)
* D) Não, todas as colunas se referem a informações do texto presente no email. Nenhuma delas é um índice ou identificador

## 4. Definição da variável-alvo

A variável-alvo é a coluna que o modelo tentará prever.

Exemplos:

- Em spam/phishing: coluna que indica se o e-mail é spam ou não;
- Em preço de casas: coluna com o preço do imóvel;
- Em vendas: coluna com o valor de vendas;
- Em depressão: coluna que indica presença ou risco de depressão.

In [ ]:
# Coloque aqui o nome da variável-alvo
# Exemplo: target = 'price'

target = 'label'

if target == '':
    print('Defina o nome da variável-alvo na variável target.')
else:
    print('Variável-alvo:', target)
    display(df[target].head())

Variável-alvo: label


,label
0,0
1,0
2,0
3,0
4,0


### Tipo de problema

Marque o tipo de problema:

- [X] Classificação binária
- [ ] Classificação multiclasse
- [ ] Regressão

Explique o motivo:

> Nosso objetivo vai ser classificar o email em uma de DUAS classes. Ou o email é phishing (1) ou não é phishing (0).

## 5. Análise de valores ausentes

Aqui vamos verificar se existem valores faltantes na base.

In [ ]:
# Quantidade de valores ausentes por coluna
missing = df.isnull().sum().sort_values(ascending=False)
missing_percent = (df.isnull().mean() * 100).sort_values(ascending=False)

missing_table = pd.DataFrame({
    'valores_ausentes': missing,
    'percentual_%': missing_percent.round(2)
})

missing_table[missing_table['valores_ausentes'] > 0]
# Printando a tabela para ficar mais visível o fato de que não há valores nulos nas colunas
missing_table

,valores_ausentes,percentual_%
num_words,0,0.0
num_unique_words,0,0.0
num_stopwords,0,0.0
num_links,0,0.0
num_unique_domains,0,0.0
num_email_addresses,0,0.0
num_spelling_errors,0,0.0
num_urgent_keywords,0,0.0
label,0,0.0


In [ ]:
# Gráfico de valores ausentes, se existirem
missing_nonzero = missing_table[missing_table['valores_ausentes'] > 0]

if len(missing_nonzero) > 0:
    missing_nonzero['percentual_%'].plot(kind='bar', figsize=(12, 5))
    plt.title('Percentual de valores ausentes por coluna')
    plt.ylabel('Percentual (%)')
    plt.xlabel('Colunas')
    plt.xticks(rotation=45, ha='right')
    plt.show()
else:
    print('Não foram encontrados valores ausentes na base.')

Não foram encontrados valores ausentes na base.


### Interpretação dos valores ausentes

Responda:

- Existem valores ausentes?
- Em quais colunas?
- Isso pode atrapalhar o treinamento?
- O grupo pretende remover linhas, preencher valores ou analisar melhor antes?

**Resposta do grupo:**

Não há valores ausentes/nulos na nossa base. A ausência de valores poderia afetar o treinamento do modelo, mas não é o nosso caso.

## 6. Estatísticas descritivas

Vamos analisar as variáveis numéricas.

In [ ]:
# Estatísticas das colunas numéricas
df.describe().T

,count,mean,std,min,25%,50%,75%,max
num_words,524846.0,276.228000,3335.456750,0.0,47.0,120.0,269.0,2339682.0
num_unique_words,524846.0,123.012167,201.626478,0.0,38.0,79.0,145.0,51251.0
num_stopwords,524846.0,80.045465,1023.330380,0.0,12.0,34.0,79.0,720411.0
num_links,524846.0,0.895588,5.603001,0.0,0.0,0.0,0.0,824.0
num_unique_domains,524846.0,0.347767,1.774209,0.0,0.0,0.0,0.0,524.0
num_email_addresses,524846.0,2.114897,13.592682,0.0,0.0,0.0,1.0,1150.0
num_spelling_errors,524846.0,24.694731,311.312358,0.0,2.0,8.0,22.0,190104.0
num_urgent_keywords,524846.0,0.245301,0.559320,0.0,0.0,0.0,0.0,7.0
label,524846.0,0.013240,0.114301,0.0,0.0,0.0,0.0,1.0


In [ ]:
# Separar colunas numéricas e categóricas
num_cols = df.select_dtypes(include=np.number).columns.tolist()
cat_cols = df.select_dtypes(exclude=np.number).columns.tolist()

print('Colunas numéricas:', num_cols)
print('\nColunas categóricas:', cat_cols)

Colunas numéricas: ['num_words', 'num_unique_words', 'num_stopwords', 'num_links', 'num_unique_domains', 'num_email_addresses', 'num_spelling_errors', 'num_urgent_keywords', 'label']

Colunas categóricas: []


### Interpretação das estatísticas

Responda:

- A) Existem variáveis com valores muito altos ou muito baixos?
- B) Alguma coluna parece ter outliers?
- C) Alguma variável numérica parece precisar de normalização?

**Resposta do grupo:**

* A) Sim. A variável "num_words", por exemplo, possui valores que variam de 0 a 2.339.682 palavras em um email. Essa faixa de valores difere bastante de outras variáveis como "num_urgent_words", que aparecem no mínimo 0 e no máximo 7 vezes em um email.
* B) Sim, é possível notar a presença de outliers em várias colunas do dataset. Na coluna "num_links", por exemplo, podemos ver que a maioria (1, 2 e 3 quartis) dos emails possuem 0 links e existe um email que possui 824 link em seu texto, valor que distoa muito dos valores padrão. Além disso, a coluna "num_spelling_errors" apresenta uma média de 8 erros por email, e existe uma instância que aparece com 190.104 erros, número totalmente fora do normal.
* C) Sim, podemos notar que quase todas as features estão em ranges muito diferentes. Por isso, precisaremos normalizá-las e colocá-las na mesma escala de valores para que seja possível efetuar um bom treinamento do modelo posteriormente.

